staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto

```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
 
 message ReopenContainerLogResponse{
 }
+
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
+    // Keep temporary files. Like log files. Helpful for debugging.
+    bool keep = 1;
+    // Checkpoint/Restore the container with established TCP connections.
+    bool tcp_established = 2;
+    // Location of the checkpoint archive used for export/import
+    string archive = 3;
+    // Specify the checkpoint archive compression algorithm
+    // See github.com/containers/storage/pkg/archive/archive.go for
+    // possible values.
+    int64 compression = 4;
+}
+
+message CheckpointContainerOptions {
+    CheckpointRestoreOptions common_options = 1;
+    // Leave the container running after checkpointing
+    bool leave_running = 2;
+}
+
+message CheckpointContainerRequest {
+    // ID of the container or pod to be checkpointed.
```

Comment:
Should we have separate requests for pod vs. containers? 
Pods may be constructed differently e.g. they could be pinned namespaces or pause (infra) containers or VMs.

Reply:
> Should we have separate requests for pod vs. containers?

From what I have tested so far with pods and containers it works well passing the ID (container or pod) to CRI-O. What I am doing in CRI-O is trying the ID first as a container ID and then as a POD ID (or the other way around) and if one matches CRI-O will checkpoint the container or pod.

> Pods may be constructed differently e.g. they could be pinned namespaces or pause (infra) containers or VMs.

Yes. The resulting checkpoints are different but CRI-O already knows how to checkpoint a container in contrast to a pod. From my point of view it can be done using a single parameter. We can also split it up in a container id and pod id parameter. It did not seem necessary so far.

Reply:
> The container id should be restricted to just container ids, overloading container ids and pod ids would require api changes.

Sorry, I also do not understand this comment. Currently I tell CRI-O to checkpoint the thing with the given ID and CRI-O figures out if it is a pod or a container.

Also, why would it require API changes? This PR introduces the API. Sorry, but I do not understand what you exactly mean. If it should be split, that is no problem, I just do not understand the reason. I guess.

Reply:
There is no rule in CRI that requires pod and container ids to be drawn from the same bucket and have a status associated to indicate whether the id is a pod or container. If you check the CRI apis you will see they either receive a pod id or a container id, never a pod or container id. 

The CRI api supports shortcutting ids, via a lookup. In your example, container ids b0c15999fad13 
052ff0835b48d, in pod 7bc90f445a4bc... I can stop the pod using crictl stopp 7, i can stop the first container using stop b.  I can't stop the pod using stop 7, or stop 7bc90f445a4bc.

Reply:
I see. Would you recommend to have completely different CRI API calls to do checkpointing and restoring?

I have the following RPC call in my PR:
```
rpc CheckpointContainer(CheckpointContainerRequest) returns (CheckpointContainerResponse) {}
```

Should that handle pod checkpointing in the future using an additional parameter? Should I rename it to have a more generic name so that it makes more sense in the future to include pod checkpointing? Should I just leave it as it is and use another RPC call in the future if we ever add pod checkpointing?

Reply:
Further .. though the odds are immense there could be a point in time where there is a pod id and a container id that have the same value yet are unique because one is a pod and the other is a container, if that is not possible, or if it should not be possible, we probably need to specify that somewhere.